xbmc/video/VideoUtils.cpp

```diff
@@ -0,0 +1,387 @@
+/*
+ *  Copyright (C) 2022 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "VideoUtils.h"
```

Comment:
This is yet another one similar to musicutils (more a GUI utils class than a generic one). The path is not explicit though which will only increase confusion. 
Can you create this in XBMC/video/GUI instead?

Reply:
Same answer as for MusicUtils - can be improved, but not by moving *as is* to a different folder. To solve this, is something I wrote on my todo list and then it will be solved in the same way for both Music and Video. I hope that's okay for you.

Reply:
Fine to me as long as it is actually improved later. Otherwise we can talk about headless as much as we want, it'll never happen :)

Reply:
And btw, isn't the **whole** "video" (and "music") folder GUI)? I cannot see any separation of core and GUI in the code. Strictly speaking, every usage of a`CFileItem` is GUI, as this class is derived from `CGUIListItem`which has members like `GUIIconOverlay`, `CGUIListItemLayout`, `ArtMap`, ... and those `CFIleItem`s are even used in the Videodatabase class. And honestly, I'm not able (skill-wise) to start separation of core and GUI for video and music component. I did that for PVR, but this is code I know very well. ;-)

Reply:
> Otherwise we can talk about headless as much as we want, it'll never happen :)

 For this particular (and the musicutils) case, I took really care that although some GUI classes are used, the code should work even without a GUI. This is why I change din the latest musicutils PR from whitelisting windows (code only works if a certain window is active) to blacklisting certain windows. Sorting items according a GUI viestate for certain windows does not rely on the windows actually being active, ether are sensible defaults. Just to give you two examples.

Reply:
At the moment everything is GUI, this is not the main point imho. There's no clear separation boundary between GUI/non-gui and personally I feel most of the decision criteria falls into common sense.  By GUI I think we actually mean having dependencies on the window system and on rendering/skin/etc. Music is something you should be able to play  even if Kodi is running without a main window (you need some user interface but not the kodi's provided one). You could use the web interface, some remote client, etc. Same for video although in that case it certainly becomes trickier.

We use CFileItem for everything just as a container to pass stuff around (and as a basis for VFS). I don't think we will be ever able to change this in the foreseeable future...
It extends CGUIFileItem but is that class actually "GUI"? It just adds a few more members that are useful regardless of GUI or headless.
 You will always need to display an icon or other artwork or to convey properties on items even if you don't have a window..
We have a lot of other examples...is GUIPassword a GUI class? Are (the low level / base classes of) context menus GUI elements?

My request here is not perfect. The logic you have in both MusicUtils or VideoUtils is not black and white when it comes to GUI/headless. But regardless... it has strong dependencies on dialogs and the window system that implies that in the current state it should be only included from GUI components. Moving it to a GUI folder would turn this implicit. For instance, If at some point in the future someone decides to include this is the Jsonrpc for to allow/expose some other use case, it's then clear that the particular usage is wrong and that a refactor is needed for that particular PR to go in. This would make us better prepared for the future than the point we are today.

Reply:
> It extends CGUIFileItem but is that class actually "GUI"? It just adds a few more members that are useful regardless of GUI or headless.

This is exactly my point for the changes I made in this and the musicutils PR. As I said, I really took care that features will work even if there is no GUI actually running.

Reply:
> Excuse me, but this is actually not true. One has to look very close of which kind those dependencies are, namely the functional runtime dependencies. Looking at compile time dependencies is not sufficient / not the right decision metrics.

ofc this is usually a too simplistic approach. But in this particular case you're including the window manager as a dependency and binding state (if a given window is active) to a particular behaviour in an utils class. Not sure how you would plan to use this in a non-gui environment without ifdefs and similar approaches.

Reply:
To be fair, I'd question whether a 'context menu' actually makes sense outside of the GUI environment?  Remotes & the web interface have their own ways of handling things so the only place you're likely to call up the context menu (or at least, these GUI related ones) is indeed in the GUI.
Headless would be fine using the jsonrpc interface as none of the stuff in this PR would be called.  Issue there is that Kodi still requires the GUI to fire up even though most functionality can be done via json - or at least, I think that's the case for the music library.